lib/trivial.nix

```
@@ -232,6 +232,31 @@ in {
 
   nixpkgsVersion = builtins.trace "`lib.nixpkgsVersion` is deprecated, use `lib.version` instead!" version;
 
+  /* The store path to this nixpkgs as a string with context. This can be used
+     to incur a dependency on this nixpkgs, for example, to symlink it into a
+     NixOS closure.
+
+     This is different than pkgs.path if nixpkgs is already in the Nix store,
+     since pkgs.path is a path to nixpkgs rather than a string-with-context,
+     and thus its usage typically, undesirably, results in copying this
+     nixpkgs.
+
+     See: https://github.com/NixOS/nix/issues/5868
+          https://github.com/NixOS/nix/issues/9428
+  */
+  nixpkgsStorePathString = if lib.trivial.inPureEvalMode || !(lib.strings.isStorePath (toString ../.))
+    # If we are in pure eval mode and we are not a flake, we cannot generate
+    # the string with context. Not much we can do.
+    then "${../.}"
+    else builtins.storePath ../.;
```

I had a hard time making sense of the logic here, I think this is better:
```suggestion
  nixpkgsStorePathString =
    if lib.strings.isStorePath (toString ../.) then
      # While `toString ../.` is already a store path here, it doesn't contain a dependency on that store path.
      # We can use `builtins.storePath` to establishes a dependency on an existing store path
      if lib.trivial.inPureEvalMode then
        # However, in pure eval mode, that builtin is not currently supported, see https://github.com/NixOS/nix/issues/5868.
        # So we fall back to string interpolation of a path, which copies the path that's already in the store to the store once more,
        # which is slow and gives it a doubly-hashed base name, but there's not much we can do
        # See also https://github.com/NixOS/nix/issues/9428
        "${../.}"
      else
        # Only outside pure eval we can actually use builtins.storePath, which can re-use the existing store path
        builtins.storePath ../.
    else
      # Here ../. is not in the store already, so we need to use string interpolation to copy the path to the store.
      "${../.}"
```